### PR TITLE
ARROW-6935: [Java] Improve the performance of comparing two blocks of heap data

### PR DESCRIPTION
**Pull request by liyafan82**
Implement methods to compare data word by word, instead of byte by byte.
Benchmarks shows that there is a 4.5x performance improvement:

ByteFunctionHelpersBenchmarks.builtInByteArrayEquals avgt 5 437.504 ± 1.120 ns/op
ByteFunctionHelpersBenchmarks.byteArrayEquals avgt 5 97.700 ± 0.178 ns/op

### CODE REVIEW

**Comment by TheNeuralBit** [comment]
I'm surprised this yielded such a big improvement, I would think `Array.equals` would be vectorized.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightDescriptor.java [hunk 1]
```diff
@@ -160,7 +161,8 @@ public boolean equals(Object obj) {
       if (other.cmd != null) {
         return false;
       }
-    } else if (!Arrays.equals(cmd, other.cmd)) {
+    } else if (ByteFunctionHelpers.equal(
```
Comment (jacques-n) [inline java/flight/src/main/java/org/apache/arrow/flight/FlightDescriptor.java]: This isn't performance sensitive. Please don't use non-idiomatic code for performance purposes where performance doesn't matter.

**Comment by liyafan82** [comment]
> I'm surprised this yielded such a big improvement, I would think `Array.equals` would be vectorized.

@TheNeuralBit Thanks for your comment.
The benchmarks ran on a machine with vectorization capability:

Intel(R) Xeon(R) CPU E5-2682 v4 @ 2.50GHz

However, Java has only limited support for SIMD. For this particular case, processing of different data is not independent (note that there is a if branch in the for loop of Arrays.equals), so JIT does not vectorize it.

**Review by jacques-n — COMMENTED** [review]
I'm struggling with this change. It seems like an optimization for a piece of code which isn't expected to be done often. In general, these heap structures aren't main path for algorithms. Why are you making this change?

File: java/flight/src/main/java/org/apache/arrow/flight/FlightDescriptor.java (hunk 1 above)
Comment (liyafan82) [reply java/flight/src/main/java/org/apache/arrow/flight/FlightDescriptor.java]: Reverted. Thanks.

**Comment by liyafan82** [comment]
> I'm struggling with this change. It seems like an optimization for a piece of code which isn't expected to be done often. In general, these heap structures aren't main path for algorithms. Why are you making this change?

I agree with you that the optimization is not on the main path of Arrow. 
There is one scenario that makes use of this optimization: the value of a VarCharVector can be loaded as a Text object, and comparing Text objects involves this optimization.

**Comment by TheNeuralBit** [comment]
> However, Java has only limited support for SIMD. For this particular case, processing of different data is not independent (note that there is a if branch in the for loop of Arrays.equals), so JIT does not vectorize it.

What java version are you using? It looks like OpenJDK ends up using some hand-written vectorized assembly in jdk 9 ([`vectorizedMismatch`](http://hg.openjdk.java.net/jdk9/jdk9/jdk/file/1edd10226c97/src/java.base/share/classes/java/util/ArraysSupport.java#l280)) but Array.equals is just a simple loop that JIT can't vectorize in jdk 8 ([link](https://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/Arrays.java#l2680))

**Comment by jacques-n** [comment]
If someone is using getObect to get the Text objects, they are already on the slow path...

**Comment by liyafan82** [comment]
> > However, Java has only limited support for SIMD. For this particular case, processing of different data is not independent (note that there is a if branch in the for loop of Arrays.equals), so JIT does not vectorize it.
> 
> What java version are you using? It looks like OpenJDK ends up using some hand-written vectorized assembly in jdk 9 ([`vectorizedMismatch`](http://hg.openjdk.java.net/jdk9/jdk9/jdk/file/1edd10226c97/src/java.base/share/classes/java/util/ArraysSupport.java#l280)) but Array.equals is just a simple loop that JIT can't vectorize in jdk 8 ([link](https://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/Arrays.java#l2680))

@TheNeuralBit Thanks a lot for the additional information.
I tried jdk 9, and it really worked. The performance is even better than our improved version. 

ByteFunctionHelpersBenchmarks.builtInByteArrayEquals  avgt    5  37.814 ± 0.167  ns/op
ByteFunctionHelpersBenchmarks.byteArrayEquals         avgt    5  93.859 ± 0.361  ns/op

To make use of vectorization in our improved version, we have changed the while loop to a for loop (please see the new commit). There are some improvements, but the performance is still not as good:

ByteFunctionHelpersBenchmarks.byteArrayEquals  avgt    5  71.725 ± 1.766  ns/op

This can be explained as: vectorizing bytes is more efficient than vectorizing longs, because the latter reduces the potential parallelism?

Another problem is that not all CPUs support vectorization. 

What do you think?

**Comment by liyafan82** [comment]
> If someone is using getObect to get the Text objects, they are already on the slow path...

Agree with you. So this PR is not so important. 
Fortunately, we have found some interesting problems in it.

**Comment by TheNeuralBit** [comment]
> This can be explained as: vectorizing bytes is more efficient than vectorizing longs, because the latter reduces the potential parallelism?

There's also a larger standard deviation in your modified benchmark (1.766 vs. 0.361 and 0.167 in the others). Maybe there's a slower initial run? It would be interesting to see what happens with a larger sample size.

> Another problem is that not all CPUs support vectorization. What do you think?

True, but I don't know that it's worth maintaining another code path for that case unless someone starts doing it, I think it will be pretty rare.

**Comment by TheNeuralBit** [comment]
> If someone is using getObect to get the Text objects, they are already on the slow path...

We are also thinking about using Arrow in Beam, which has a similar computational model to Flink that is entirely element-wise. It would be easier for us to support arrow over time by first processing elements that are views into the underlying arrow buffer, and then starting to provide batched versions of expensive/common transforms. So it would be nice if the "slow path" got a little love, since we may be on it for a while.

That being said I agree that we'd be better off just relying on Java >9 and `vectorizedMismatch` for this particular path, I just don't want to discourage optimizing the element-wise path entirely.

**Comment by liyafan82** [comment]
> There's also a larger standard deviation in your modified benchmark (1.766 vs. 0.361 and 0.167 in the others). Maybe there's a slower initial run? It would be interesting to see what happens with a larger sample size.

I rerun the benchmark with a larger sample size (1024 * 1024, the original was 1024). The results are consistent with your expectation: for loop is not faster than while loop:

while loop
ByteFunctionHelpersBenchmarks.builtInByteArrayEquals  avgt    5  67789.274 ± 5436.789  ns/op
ByteFunctionHelpersBenchmarks.byteArrayEquals         avgt    5  90008.613 ±  400.723  ns/op

for loop
ByteFunctionHelpersBenchmarks.builtInByteArrayEquals  avgt    5  68210.532 ±  71.264  ns/op
ByteFunctionHelpersBenchmarks.byteArrayEquals         avgt    5  88928.010 ± 439.912  ns/op

So it seems that java 9 could not vectorized the code of our improved version. (I have not managed to see the assembly for jdk9, the hsdis-amd64.so seems not loadable).

Now we have:
1. for jdk8, our improved version is 4+x times faster
2. for jdk9, the built-in version (Arrays.equal) is 2+ times faster.

What is your suggestion?

> 
> > Another problem is that not all CPUs support vectorization. What do you think?
> 
> True, but I don't know that it's worth maintaining another code path for that case unless someone starts doing it, I think it will be pretty rare.

**Comment by liyafan82** [comment]
> > If someone is using getObect to get the Text objects, they are already on the slow path...
> 
> We are also thinking about using Arrow in Beam, which has a similar computational model to Flink that is entirely element-wise. It would be easier for us to support arrow over time by first processing elements that are views into the underlying arrow buffer, and then starting to provide batched versions of expensive/common transforms. So it would be nice if the "slow path" got a little love, since we may be on it for a while.
> 
> That being said I agree that we'd be better off just relying on Java >9 and `vectorizedMismatch` for this particular path, I just don't want to discourage optimizing the element-wise path entirely.

Cool. Thanks for sharing the good news. 
Hope you will successfully incorporate Arrow to Beam, and make Beam better.

**Comment by TheNeuralBit** [comment]
> What is your suggestion?

I agree with @jacques-n that it's not worth introducing the complexity for this improvement. Particularly when users can get a more significant speedup just by using a more recent version of Java. I think we should encourage that instead.

> I agree with you that most CPU supports SIMD. I am just curious what the performance would be like in a typical working environement for Arrow. Our environment is a high performance computer with strong SIMD capability (Intel AVX2).

I think a high performance computer with strong SIMD support is the target working environment for Arrow :) 

I think AVX and AVX2 are pretty [ubiquitous](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions#CPUs_with_AVX2) now anyway. I looked around to see if I had a machine at my disposal that doesn't support AVX or AVX2 that I could try running some benchmarks on but I can't find one, even my i3-4130T (from Q3 2013) supports both.

**Comment by liyafan82** [comment]
> > What is your suggestion?
> 
> I agree with @jacques-n that it's not worth introducing the complexity for this improvement. Particularly when users can get a more significant speedup just by using a more recent version of Java. I think we should encourage that instead.
> 
> > I agree with you that most CPU supports SIMD. I am just curious what the performance would be like in a typical working environement for Arrow. Our environment is a high performance computer with strong SIMD capability (Intel AVX2).
> 
> I think a high performance computer with strong SIMD support is the target working environment for Arrow :)
> 
> I think AVX and AVX2 are pretty [ubiquitous](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions#CPUs_with_AVX2) now anyway. I looked around to see if I had a machine at my disposal that doesn't support AVX or AVX2 that I could try running some benchmarks on but I can't find one, even my i3-4130T (from Q3 2013) supports both.

@TheNeuralBit Thanks a lot for your valuable feedback.

Given the current comments, I don't think this PR deserves to be merged. 
However, the discussions are really interesting and helpful. So I will leave it open for a few days, and if there is no more discussion, I will close it.

**Comment by liyafan82** [comment]
Closing this. Thanks a lot for your valuable comments. @TheNeuralBit @jacques-n

**Pull request closed without merging**